File: airbyte-commons-worker/src/main/java/io/airbyte/workers/context/AttemptContext.java
/*
 * Copyright (c) 2020-2025 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.workers.context;

import static io.airbyte.metrics.lib.ApmTraceConstants.Tags.ATTEMPT_NUMBER_KEY;
import static io.airbyte.metrics.lib.ApmTraceConstants.Tags.CONNECTION_ID_KEY;
import static io.airbyte.metrics.lib.ApmTraceConstants.Tags.JOB_ID_KEY;

import io.airbyte.metrics.lib.ApmTraceUtils;
import java.util.HashMap;
import java.util.Map;
import java.util.UUID;

/**
 * Context of an Attempt.
 */
public record AttemptContext(UUID connectionId, Long jobId, Integer attemptNumber) {

  /**
   * Update the current trace with the ids from the context.
   */
  public void addTagsToTrace() {
    final Map<String, Object> tags = new HashMap<>();
    if (connectionId != null) {
      tags.put(CONNECTION_ID_KEY, connectionId);
    }
    if (jobId != null) {
      tags.put(JOB_ID_KEY, jobId);
    }
    if (attemptNumber != null) {
      tags.put(ATTEMPT_NUMBER_KEY, attemptNumber);
    }
    ApmTraceUtils.addTagsToTrace(tags);
  }

}

File: airbyte-workers/src/main/kotlin/io/airbyte/workers/context/AttemptContext.kt
/*
 * Copyright (c) 2020-2025 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.workers.context

import io.airbyte.metrics.lib.ApmTraceConstants
import io.airbyte.metrics.lib.ApmTraceUtils
import java.util.UUID

data class AttemptContext(
  val connectionId: UUID?,
  val jobId: Long?,
  val attemptNumber: Int?,
) {
  /**
   * Update the current trace with the ids from the context.
   */
  fun addTagsToTrace() {
    val tags: MutableMap<String?, Any?> = mutableMapOf()
    if (connectionId != null) {
      tags.put(ApmTraceConstants.Tags.CONNECTION_ID_KEY, connectionId)
    }
    if (jobId != null) {
      tags.put(ApmTraceConstants.Tags.JOB_ID_KEY, jobId)
    }
    if (attemptNumber != null) {
      tags.put(ApmTraceConstants.Tags.ATTEMPT_NUMBER_KEY, attemptNumber)
    }
    ApmTraceUtils.addTagsToTrace(tags)
  }
}
